server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java

```diff
@@ -1186,6 +1238,586 @@ protected void pushToUpdateQueue(final List<NetworkVO> networks) throws Interrup
+        public FetchRouterHealthChecksResultTask() {
+        }
+
+        @Override
+        protected void runInContext() {
+            try {
+                final List<DomainRouterVO> routers = _routerDao.listByStateAndManagementServer(VirtualMachine.State.Running, mgmtSrvrId);
+                s_logger.info("Found " + routers.size() + " running routers. Fetching, analysing and updating DB for the health checks.");
+                if (!RouterHealthChecksEnabled.value()) {
+                    s_logger.debug("Skipping fetching of router health check results as router.health.checks.enabled is disabled");
+                    return;
+                }
+
+                for (final DomainRouterVO router : routers) {
+                    GetRouterMonitorResultsAnswer answer = fetchAndUpdateRouterHealthChecks(router, false);
+                    List<String> failingChecks = getFailingChecks(router, answer);
+                    handleFailingChecks(router, failingChecks);
+                }
+            } catch (final Exception ex) {
+                s_logger.error("Fail to complete the FetchRouterHealthChecksResultTask! ", ex);
+                ex.printStackTrace();
+            }
+        }
+
+        private List<String> getFailingChecks(DomainRouterVO router, GetRouterMonitorResultsAnswer answer) {
+            if (answer == null) {
+                s_logger.warn("Unable to fetch monitor results for router " + router);
+                updateRouterConnectivityHealthCheck(router.getId(), false, "Communication failed");
+                return Arrays.asList(CONNECTIVITY_TEST);
+            } else if (!answer.getResult()) {
+                s_logger.warn("Failed to fetch monitor results from router " + router + " with details: " + answer.getDetails());
+                updateRouterConnectivityHealthCheck(router.getId(), false, "Failed to fetch results with details: " + answer.getDetails());
+                return Arrays.asList(CONNECTIVITY_TEST);
+            } else {
+                updateRouterConnectivityHealthCheck(router.getId(), true, "Successfully fetched data");
+                updateDbHealthChecksFromRouterResponse(router.getId(), answer.getMonitoringResults());
+                return answer.getFailingChecks();
+            }
+        }
+
+        private void handleFailingChecks(DomainRouterVO router, List<String> failingChecks) {
+            if (failingChecks == null || failingChecks.size() == 0) {
+                return;
+            }
+
+            String alertMessage = "Health checks failed: " + failingChecks.size() + " failing checks on router " + router.getUuid();
+            _alertMgr.sendAlert(AlertType.ALERT_TYPE_DOMAIN_ROUTER, router.getDataCenterId(), router.getPodIdToDeployIn(),
+                    alertMessage, alertMessage);
+            s_logger.warn(alertMessage + ". Checking failed health checks to see if router needs reboot");
```

Comment:
Change to "if the router needs to be recreated" (i.e. not rebooted)

Reply:
Done

Reply:
Not sure it's done?